src/main/java/org/fcrepo/client/BodyRequestBuilder.java

```diff
@@ -0,0 +1,144 @@
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+
+import org.apache.http.client.methods.HttpEntityEnclosingRequestBase;
+import org.apache.http.client.methods.HttpRequestBase;
+import org.apache.http.entity.InputStreamEntity;
+import org.slf4j.Logger;
+
+/**
+ * Request builder which includes a body component
+ * 
+ * @author bbpennel
+ */
+public abstract class BodyRequestBuilder<T extends BodyRequestBuilder<T>> extends
+        RequestBuilder<BodyRequestBuilder<T>> {
+
+    private static final Logger LOGGER = getLogger(PatchBuilder.class);
+
+    protected InputStream bodyStream;
+
+    protected String contentType;
+
+    protected String digest;
+
+    protected String etag;
+
+    protected String unmodifiedSince;
+
+    /**
+     * Instantiate builder
+     * 
+     * @param uri uri request will be issued to
+     * @param client the client
+     */
+    protected BodyRequestBuilder(final URI uri, final FcrepoClient client) {
+        super(uri, client);
+    }
+
+    protected abstract T self();
+
+    @Override
+    protected void populateRequest(final HttpRequestBase request) throws FcrepoOperationFailedException {
+        addBody((HttpEntityEnclosingRequestBase) request);
+
+        addIfUnmodifiedSince(request);
+        addIfMatch(request);
+
+        addDigest(request);
+
+        LOGGER.debug("Fcrepo {} request headers: {}", request.getMethod(), (Object[]) request.getAllHeaders());
+    }
+
+    /**
+     * Add a body to this request from a stream, with application/octet-stream as its content type
+     * 
+     * @param stream InputStream of the content to be sent to the server
+     * @return this builder
+     */
+    public T body(final InputStream stream) {
+        return body(stream, null);
+    }
+
+    /**
+     * Add a body to this request as a stream with the given content type
+     * 
+     * @param stream InputStream of the content to be sent to the server
+     * @param contentType the Content-Type of the body
+     * @return this builder
+     */
+    public T body(final InputStream stream, final String contentType) {
+        this.bodyStream = stream;
+        if (contentType == null) {
+            this.contentType = "application/octet-stream";
+        } else {
+            this.contentType = contentType;
+        }
+
+        return self();
+    }
+
+    /**
+     * Add the given file as the body for this request with the provided content type
+     * 
+     * @param file File containing the content to be sent to the server
+     * @param contentType the Content-Type of the body
+     * @return this builder
+     * @throws IOException when unable to stream the body file
+     */
+    public T body(final File file, final String contentType) throws IOException {
+        return body(new FileInputStream(file), contentType);
+    }
+
+    protected void addBody(final HttpEntityEnclosingRequestBase request) {
```

Comment:
After considering a couple ways of trying to make this fluent, I wonder if it would make more sense for the RequestBuilder objects to immediately instantiate a HttpRequestBase rather than requiring the use of addBlah methods later, and remove populateRequest:

```java
public class PostBuilder... {
    protected HttpRequestBuilder request;

    public PostBuilder(...) {
        super(...);
        this.request = HttpMethods.POST.createRequest(targetUri);
    }

    public PostBuilder<T> slug(final String slug) {
        if (slug != null) {
            request.addHeader(SLUG, slug);
        }
        return self();
    }
}

public abstract class BodyRequestBuilder... {
    public T body(final InputStream stream, final String contentType) {
        if (bodyStream != null) {
            request.setEntity(new InputStreamEntity(bodyStream));
            request.addHeader(CONTENT_TYPE, contentType);
        }
        return self();
    }
```